Builds - Mount trusted CA for cluster proxies

gabemontero opened the pull request:
@openshift/openshift-team-developer-experience fyi

/assign @adambkaplan 
/assign @bparees

gabemontero commented:
/test e2e-aws-builds

gabemontero commented:
but PTAL @adambkaplan @bparees even with WIP label

gabemontero commented:
hmmm ....  can we not run e2e-aws-builds out of OCM ?

gabemontero commented:
we cannot:   https://github.com/openshift/release/blob/master/ci-operator/config/openshift/openshift-controller-manager/openshift-openshift-controller-manager-master.yaml

I'll be submitting a openshift/release PR to fix that @adambkaplan @bparees

gabemontero commented:
e2e-aws errors were oauth/router flakes

/test e2e-aws

adambkaplan commented on pkg/build/controller/strategy/util.go:
```diff
@@ -32,6 +32,10 @@ const (
+	//TODO Ben and Adam were discussing this in https://github.com/openshift/cluster-image-registry-operator/pull/340
+	//with Adam noting he asked for clarification in the proxy design doc; a little curious if this should be a
+	//directory to hold multiple pem files, or will this contain an aggregated item
+	ConfigMapBuildProxyCAMountPath = "/etc/pki/ca-trust/extracted/pem/tls-ca-bundle.pem"
```
Per arch call this is the correct mount path, but we should mount to a directory (not an individual file).

```suggestion
	ConfigMapBuildProxyCAMountPath = "/etc/pki/ca-trust/extracted/pem/"
```

gabemontero replied:
part of next push

adambkaplan commented on pkg/build/controller/strategy/util.go:
```diff
@@ -531,6 +538,45 @@ func setupBuildCAs(build *buildv1.Build, pod *corev1.Pod, additionalCAs map[stri
+			LocalObjectReference: corev1.LocalObjectReference{
+				Name: buildutil.GetBuildProxyCAConfigMapName(build),
+			},
+			//TODO the TBD proxy CA injector will update the ConfigMapVolumeSource keyToPath items, right?
```
Correct - once the work for the controller is done, we need to explicitly reference the injected key here and give it the new file name `tls-ca-bundle.pem`.

gabemontero replied:
If the key name is defined yet in the design doc, I missed it.  Do we know that and the other relevant pieces to scaffold this (even if we comment it out in the short term) @adambkaplan ?

adambkaplan replied:
As of right now the injected key will be `ca-bundle.crt` - it's buried and easy to miss.

I made a note that the behavior - once finalized, needs to be made abundantly clear:
https://docs.google.com/a/redhat.com/document/d/1y0t0yEOSnKc4abxsjxEQjrFa1AP8iHcGyxlBpqGLO08/edit?disco=AAAADYn95ZA

gabemontero replied:
ok I'm going to add some scaffold/prototype code in the commented out section to help capture this

adambkaplan commented on pkg/build/controller/build/build_controller.go:
```diff
@@ -1234,6 +1252,12 @@ func (bc *BuildController) createBuildPod(build *buildv1.Build) (*buildUpdate, e
+			if err != nil {
+				return update, err
+			}
+		}
```
@bparees per arch call my understanding is that we will always want to create this ConfigMap, and that ca bundle will contain all trust for the cluster.

For the sake of staging code these lines should be commented out until the new controller that injects the trusted CA bundle lands.

gabemontero replied:
OK so remove the "is proxy configured" qualifiers once @bparees confirms @adambkaplan 's understanding from the arch call

gabemontero replied:
commenting out part of next push

bparees replied:
you should always create the TrustedCA configmap, it should be annotated for injection, and you should always mount it into your pod.  In some cases this will mean that you mount an empty/non-existent trust bundle into your pod.  This is considered desirable behavior.

and yes, until the trusted CA controller exists, you do not wan to actually mount CM.

adambkaplan replied:
@bparees we should clarify - we want to explicitly reference the injected key so that the build pod doesn't start until the trusted CA controller adds the ca-bundle, as we do with the service-ca bundle for the registry.

bparees replied:
sorry, for build-pods you're correct.  they are a bit of a special case.

gabemontero replied:
see https://github.com/openshift/openshift-controller-manager/pull/12#discussion_r309415769 I plan to get some code along those lines in the commented out section to help capture this

adambkaplan commented on pkg/build/controller/build/build_controller.go:
```diff
@@ -1210,6 +1216,18 @@ func (bc *BuildController) createBuildPod(build *buildv1.Build) (*buildUpdate, e
+					return update, err
+				}
+			}
+		}
```
Ditto - this should be commented out until the new trusted CA controller lands.

gabemontero replied:
part of next push

gabemontero commented:
@adambkaplan I pushed some responses to your initial comments in a new commit

I deferred for now on the "always create the config map" point since it seemed like you were waiting for confirmation from @bparees

gabemontero commented:
OK @adambkaplan @bparees always created the CA

relevant code snippets commented out pending arrival of injector controller

squashed commits 

renamed `ProxyCA` to `GlobalCA` and the like 

I have the one question on potential additional scaffolding that could be done based whether injection key names are known.

PTAL

gabemontero commented:
and the e2e-aws-builds jobs is getting triggered now

gabemontero commented on pkg/build/buildutil/util.go:
```diff
@@ -36,9 +36,13 @@ const (
+	globalCAConfigMapSuffix = "global-ca"
+	// GlobalCAConfigMapAnnotation is the annotation key to set on a config map so that the platform's
+	// proxy support will inject any required CA's needed for proxy communication.
+	GlobalCAConfigMapAnnotation = "config.openshift.io/inject-proxy-cabundle"
```
I did not change `inject-proxy-cabundle` because I did not see that the annotation name had changed yet in the design doc

adambkaplan replied:
Makes sense - I'd expect there will be more churn here, but keep it for now. Design is _evolving_ to say the least.

adambkaplan commented:
/approve

I'd like to see e2e-aws-builds succeed before staging this

gabemontero commented:
e2e-aws-build passed

gabemontero commented:
/hold

temporarily until I get some of the inject key reference prototype in the commented out section

gabemontero commented:
ok placeholder added in commented out section

/hold cancel

gabemontero commented:
terraform flake on e2e-aws

/test e2e-aws

gabemontero commented:
/hold

they renamed the annotation per @adambkaplan

gabemontero commented:
/hold cancel

annotation updated @adambkaplan please re-lgtm

gabemontero commented:
auth / mem related flakes e2e-aws-builds

/test e2e-aws-builds

gabemontero commented:
`failed to acquire resource` flakes discussed in forum-testplatform

/test e2e-aws-builds

@adambkaplan please re-post the lgtm thx

gabemontero commented:
e2e-aws-build avoided flakes ... this is ripe for the re posting the lgtm @adambkaplan

adambkaplan commented:
/lgtm
